### PR TITLE
[e2e] fix race condition workspace

### PR DESCRIPTION
**Pull request by pducolin**
### What does this PR do?

Create one workspace per stack, to avoid race conditions on common `Pulumi.yaml`. 

### Motivation

We currently use one common Pulumi workspace for all tests. Each test defines one Pulumi stack, and each Pulumi stack has its own Pulumi program, aka the run function.

The problem is a Pulumi workspace supports multiple stacks, but one single program across all stacks

> To enable a broad range of runtime customization the API defines a Workspace interface. A Workspace is the execution context containing a single Pulumi project, a program, and multiple stacks. Workspaces are used to manage the execution environment, providing various utilities such as plugin installation, environment configuration ($PULUMI_HOME), and creation, deletion, and listing of Stacks.

Source: https://www.pulumi.com/docs/using-pulumi/automation-api/concepts-terminology/#workspace

We recently changed the lock logic around stacks, allowing running multiple stacks in parallel. This has the advantage of using the same runner to run multiple e2e tests in parallel. But it also raised flakiness on some e2e tests jobs, running multiple stacks in parallel.

Here is an example of failing job: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/361860810#L503

The error is reported when accessing `/tmp/Pulumi.yaml`. While each stack has its own `Pulumi.<stack>.yaml`, there is one single `/tmp/Pulumi.yaml` per workspace. If multiple stacks access this file, they break code execution causing a race condition in accessing and modifying the workspace file.

For this reason, we now have one workspace per stack. 

### Possible Drawbacks / Trade-offs

This change has the disadvantage of requiring one extra step to list all existing stacks on a host. The cleanup task was updated and tested to support cleaning both legacy and new implementation.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by KevinFairise2 — APPROVED** [review]
Nice! 🚀

File: test/new-e2e/pkg/runner/ci_profile.go [hunk 1]
```diff
@@ -73,9 +70,11 @@ func NewCIProfile() (Profile, error) {
 	}, nil
 }
 
-// RootWorkspacePath returns the root directory for CI Pulumi workspace
-func (p ciProfile) RootWorkspacePath() string {
-	return workspaceFolder
+// GetWorkspacePath returns the directory for CI Pulumi workspace.
+// Since one Workspace supports one single program and we have one program per stack,
+// the path should be unique for each stack.
+func (p ciProfile) GetWorkspacePath(stackName string) string {
```
Comment (vboulineau) [inline test/new-e2e/pkg/runner/ci_profile.go:76]: I think it can be moved to base

File: test/new-e2e/pkg/runner/ci_profile.go [hunk 2]
```diff
@@ -73,9 +70,11 @@ func NewCIProfile() (Profile, error) {
...
+	return path.Join(workspaceRootFolder, stackName)
```
Comment (vboulineau) [inline test/new-e2e/pkg/runner/ci_profile.go:77]: You may want to sanitize or hash `stackName` to avoid weird paths (esp. on Windows).

Comment (pducolin) [reply test/new-e2e/pkg/runner/ci_profile.go:77]: Good catch !

File: tasks/new_e2e_tests.py [hunk 3]
```diff
@@ -201,47 +205,55 @@ def _clean_stacks(ctx: Context):
...
+    workspace_subdirs_root = os.path.join(tempfile.gettempdir(), "pulumi-workspace")
+    if os.path.isdir(workspace_subdirs_root):
+        for entry in os.listdir(Path(workspace_subdirs_root)):
+            subdir = os.path.join(workspace_subdirs_root, entry)
+            if os.path.isdir(subdir):
+                workspace_dirs.append(subdir)
+    e2e_stacks: List[Stack] = []
+    for workspace_dir in workspace_dirs:
+        with ctx.cd(workspace_dir):
+            output = ctx.run("PULUMI_SKIP_UPDATE_CHECK=true pulumi stack ls", pty=True)
```
Comment (vboulineau) [inline tasks/new_e2e_tests.py:221]: What about using structured output instead `pulumi stack ls --all --json`?

**Review by vboulineau — APPROVED** [review]

**Pull request merged**